File: src/back-end/FunChess/src/FunChess.Core.Chess/Pieces/Knight.cs
using FunChess.Core.Chess.Enums;
using FunChess.Core.Chess.Structs;

namespace FunChess.Core.Chess.Pieces;

internal sealed class Knight : Piece
{
    public override bool Move(Board board, Move move)
    {
        if (!MoveIsValid(board, move, out _)) return false;

        board.InternalBoard[move.Next.Index] = board.InternalBoard[move.Previous.Index];
        board.InternalBoard[move.Previous.Index] = Cell.Empty;

        return true;
    }
    
    public override bool MoveIsValid(Board board, Move move, out SpecialMove specialMove)
    {
        specialMove = SpecialMove.None;
        if (move.DiffX is 1 or -1)
        {
            return move.DiffY is 2 or -2;
        }
        if (move.DiffY is 1 or -1)
        {
            return move.DiffX is 2 or -2;
        }
        return false;
    }
}

File: src/front-end/fun-chess-app/src/core/chess/pieces/Knight.ts
import Piece from "../Piece";
import Board from "../Board";
import SpecialMove from "../enums/SpecialMove";
import Move from "../structs/Move";

class Knight extends Piece {
    private constructor() {
        super(32);
    }

    public static readonly instance: Piece = new Knight();

    public override moveIsValid(_board: Board, move: Move): [boolean, SpecialMove] {
        const specialMove = SpecialMove.None;
        if (move.diffX === 1 || move.diffX === -1) {
            return [move.diffY === 2 || move.diffY === -2, specialMove];
        }
        if (move.diffY === 1 || move.diffY === -1) {
            return [move.diffX === 2 || move.diffX === -2, specialMove];
        }
        return [false, specialMove];
    }
}

export default Knight;